### ISSUE TITLE
Crashes on ill-formatted user input

### ISSUE THREAD

**Issue by zzril** [hrushikeshrv/chessengine#9]
Currently, if we input a move in wrong format (e. g. "2. d4" or "d2-d4" instead of "d4"), the program crashes.

Better would be to catch the error that is thrown and give the user a chance to re-type it. (Basically, asking for user input in a loop, repeating until the move can be parsed successfully.)

---

Example input and crash:  
```
Enter the move you want to make in standard algebraic notation - d2-d4
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/kali/repos/chessengine/chessengine/play.py", line 15, in <module>
    b.play()
  File "/home/kali/repos/chessengine/chessengine/bitboard.py", line 672, in play
    self.move_san(move=move, side=side_to_move)
  File "/home/kali/repos/chessengine/chessengine/bitboard.py", line 438, in move_san
    raise ValueError(f"{move} is not a valid move for {side}.")
ValueError: d2-d4 is not a valid move for white.
```

**Comment by Substancia** [comment]
I had asked if I could take up the "dark theme piece colour" issue. If this issue isn't taken up by the time I get to the dark theme issue, can I take this up? If anyone else is interested in this one before that, happy to comply :)

**Comment by hrushikeshrv** [comment]
I should have done this the first time I implemented the game loop haha! @Substancia could you take up #10 and @zzril could take up this issue?

**Comment by hrushikeshrv** [comment]
As an enhancement, we could also allow users to enter input at 2 squares separated by a delimiter. For example, users should be able to enter their moves as `d2-d4` (as you did in your example), and that would mean moving the piece on `d2` to `d4`. Checking if that is a valid move is already done by the engine, so we will only need to add the logic required to do this in `chessengine.bitboard.Board.play` to check the input format. This would be helpful for users who are not familiar with SAN.

**Comment by zzril** [comment]
> I should have done this the first time I implemented the game loop haha! @Substancia could you take up #10 and @zzril could take up this issue?

Yes, I can work on this.

**Comment by zzril** [comment]
Fixed this. See Pull Request #24 .

On a side note, I think the `play` method in `bitboard.py` could benefit from some refactoring, especially by extracting code into subroutines. We might want to raise an own issue for that. Need to time it well, however, so that it doesn't produce a bunch of merge conflicts due to others working on that method at the same time. ;)

**Comment by zzril** [comment]
> As an enhancement, we could also allow users to enter input at 2 squares separated by a delimiter. For example, users should be able to enter their moves as `d2-d4` (as you did in your example), and that would mean moving the piece on `d2` to `d4`. Checking if that is a valid move is already done by the engine, so we will only need to add the logic required to do this in `chessengine.bitboard.Board.play` to check the input format. This would be helpful for users who are not familiar with SAN.

I can work on that as well, but would prefer to handle this as a seperate issue once my current PR is merged. Otherwise, the merge conflicts it produces might become annoying. ;-)

**Comment by hrushikeshrv** [comment]
Yup, merged 👍🏻 :)

**Comment by zzril** [comment]
@hrushikeshrv Concerning the parsing of different input formats:

My current plan is to write an additional method within the `play` method in `bitboard.py` that translates user input to standard algebraic notation. Signature would be sthg like `to_san(user_input: str) -> str`.  
Then, instead of later calling `self.move_san(move=move, side=side_to_move)` directly (with `move` being the user input in this case), I would call `self.move_san(move=to_san(move), side=side_to_move)`.  
(You can also see this implementation in [the according branch in my forked repo](https://github.com/zzril/chessengine/blob/allow-different-input-formats/chessengine/bitboard.py).)  
This way, I do not have to touch the parser class and it stays more modular. In general, I would prefer this solution.

However, if you think the translation of *any* notation format to standard algebraic notation could also be needed elsewhere, I could also try to understand all the Regexes in the Parser class and try to update it there. :D But again, that class is only for PGN parsing, and according to Wiki, that means the moves will already be in SAN, right?

So, is it ok if I go with the solution shown above, translate the user input to SAN, then let the `move_san` function to its job?

**Comment by hrushikeshrv** [comment]
Yes, that's what I was thinking as well. Adding a function like `to_san` would be good, but I was thinking of making it available on the `Board` class as a documented API method, since it seems like it could be useful to (future) users who want to extend the `Board` class.

Then, if we need to add support for other notations, we could maybe add an argument to `to_san` describing the notation that it should translate from.

**Comment by zzril** [comment]
@hrushikeshrv Alright, should I define the method in `utils.py` and add to the imports in `bitboard.py` then? After all, this translation operation doesn't really have to do with the Board itself, so I think it makes sense to declare it globally, rather than as a method of the `Board` class?

(I'm thinking of sthg along the lines of [this](https://github.com/zzril/chessengine/commit/c87b274ac5ca9b9d9534dc65222544aa46aa5575).)

**Comment by hrushikeshrv** [comment]
@zzril I agree, that makes more sense. Go for it 👍🏻
